hints/hint_distribution.py

```diff
@@ -0,0 +1,461 @@
+                    num_items = len(self.hintable_items)
+                for i in range(num_items):
+                    self.hints.extend([self._create_item_hint()] * curr_type["copies"])
+            elif type == "random":
+                num_random = curr_type["fixed"]
+                for i in range(num_random):
+                    # no failsafe is needed here as there should also be progress locations to hint that have not been previously hinted
+                    self.hints.extend(
+                        [self._create_random_hint()] * curr_type["copies"]
+                    )
+            elif type == "junk":
+                self.hints.append(
+                    EmptyGossipStoneHint(None, None, False, self.junk_hints.pop())
+                )
+            else:
+                raise InvalidHintDistribution(
+                    "Invalid hint type found in distribution while geneating fixed hints"
+                )
+
+        # reverse the list of hints to we can pop off the back in O(1) in next_hint ()
+        # this also preserves the order they were added as they are removed so that order parameter is repsected
+        self.hints.reverse()
+
+        for hint_type in self.distribution.keys():
+            self.weighted_types.append(hint_type)
+            self.weights.append(self.distribution[hint_type]["weight"])
+
+        self.junk_hints = JUNK_TEXT.copy()
+        self.rng.shuffle(self.junk_hints)
+
+    """
+    Uses the distribution to calculate the next hint
+    """
+
+    def next_hint(self) -> GossipStoneHint:
+        if len(self.hints) > 0:
+            return self.hints.pop()
+        [next_type] = self.rng.choices(self.weighted_types, self.weights)
+        type = self.distribution[next_type]
+        if next_type == "sometimes":
+            hint = self._create_sometimes_hint()
+            if type["copies"]:
+                self.hints.extend([hint] * (type["copies"] - 1))
+            return hint
+        elif next_type == "sots":
+            hint = self._create_sots_hint()
+            if type["copies"]:
+                self.hints.extend([hint] * (type["copies"] - 1))
+            return hint
+        elif next_type == "barren":
+            hint = self._create_barren_hint()
+            if type["copies"]:
+                self.hints.extend([hint] * (type["copies"] - 1))
+            return hint
+        elif next_type == "item":
+            hint = self._create_item_hint()
+            if type["copies"]:
+                self.hints.extend([hint] * (type["copies"] - 1))
+            return hint
+        elif next_type == "random":
+            hint = self._create_random_hint()
+            if type["copies"]:
+                self.hints.extend([hint] * (type["copies"] - 1))
+            return hint
+        # junk hint is the last possible type and also a fallback
+        return EmptyGossipStoneHint(None, None, False, self.junk_hints.pop())
```

Comment:
```suggestion
        elif next_type == "random":
            hint = self._create_random_hint()
        else:
            # junk hint is the last possible type and also a fallback
            return EmptyGossipStoneHint(None, None, False, self.junk_hints.pop())
        if type["copies"]:
            self.hints.extend([hint] * (type["copies"] - 1))
        return hint
```
deduplicate the copy mechanism